{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8b16c7af-ec78-4f5b-a53e-18de83b29485",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import random\n",
    "import networkx as nx\n",
    "import collections\n",
    "from scipy.stats import wilcoxon, pearsonr\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "random.seed(1999)\n",
    "np.random.seed(1999)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e823c2c-1d0f-4887-a862-8f509da85a22",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/SLURM_23659463/ipykernel_25771/458702357.py:2: DtypeWarning: Columns (49,50,51,53) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mutation = pd.read_csv(\"../../dataset/OmicsSomaticMutations.csv\")\n",
      "/tmp/SLURM_23659463/ipykernel_25771/458702357.py:6: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  clinvar = pd.read_csv(\"../../dataset/ClinVar/ClinVar_variant_summary.txt\", delimiter='\\t')\n"
     ]
    }
   ],
   "source": [
    "data = pd.read_csv(\"../../dataset/CRISPRGeneEffect.csv\", index_col=0)\n",
    "mutation = pd.read_csv(\"../../dataset/OmicsSomaticMutations.csv\")\n",
    "mutation = mutation[mutation['VariantType'] == 'SNP']\n",
    "mutation = mutation[['Chrom', 'Pos', 'HugoSymbol', 'ModelID']]\n",
    "\n",
    "clinvar = pd.read_csv(\"../../dataset/ClinVar/ClinVar_variant_summary.txt\", delimiter='\\t')\n",
    "clinvar = clinvar[clinvar['Assembly'] == \"GRCh38\"]\n",
    "clinvar = clinvar[clinvar['Type'] == \"single nucleotide variant\"]\n",
    "pathogenicity = pd.read_csv(\"../../dataset/ClinVar/pathogenicity.csv\", index_col=0)\n",
    "pathogenetic_type = list(pathogenicity[pathogenicity['Pathogenicity'] == 'Y']['Category'])\n",
    "mutation_patho = clinvar[clinvar['ClinicalSignificance'].isin(pathogenetic_type)]\n",
    "mutation_patho = mutation_patho[[\"Chromosome\", \"Start\", \"GeneSymbol\"]]\n",
    "mutation_patho['Chromosome'] = mutation_patho['Chromosome'].apply(lambda x: 'chr' + str(x))\n",
    "mutation['ID'] = mutation['Chrom'] + '-' + mutation['Pos'].astype(str)\n",
    "mutation_patho['ID'] = mutation_patho['Chromosome'] + '-' + mutation_patho['Start'].astype(str)\n",
    "mutation = mutation.sort_values(by=['ID'])\n",
    "mutation_patho = mutation_patho.sort_values(by=['ID'])\n",
    "mutation = pd.merge(mutation, mutation_patho, on='ID', how='inner')\n",
    "mutation = mutation[['ModelID', 'HugoSymbol']]\n",
    "mutation = mutation.sort_values([\"HugoSymbol\", \"ModelID\"])\n",
    "mutation = mutation.drop_duplicates()\n",
    "mutation.index = range(mutation.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7a6ee3c8-cae1-4cc7-9c3d-714da3bfc6e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_col = list(data.columns)\n",
    "for i in range(len(gene_col)):\n",
    "    gene_col[i] = gene_col[i].split(' ')[0]\n",
    "data.columns = gene_col\n",
    "gene_info = pd.read_csv(\"../../dataset/InputGene/ScreenedGene.csv\")\n",
    "gene_info = gene_info.drop_duplicates(subset=['From'], keep='first')\n",
    "data = data[gene_info.iloc[:, 0]]\n",
    "data.columns = gene_info.iloc[:, 1]\n",
    "ifnull = data.isnull().sum()\n",
    "data = data[ifnull[ifnull == 0].index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fa21f916-1c3c-41eb-b369-e8a82fd26025",
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_gene = pd.read_csv(\"../../dataset/InputGene/temp.csv\", header=None)\n",
    "label = pd.DataFrame(data=0, index=mut_gene.iloc[:, 0], columns=data.index)\n",
    "for i in range(label.shape[0]):\n",
    "    mut_sub = mutation[mutation['HugoSymbol'] == mut_gene.iloc[i, 0]]\n",
    "    model_sub = list(set(mut_sub['ModelID']) & set(label.columns))\n",
    "    label.iloc[i][model_sub] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1c610350-7a39-40da-bbd3-1803c214b1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_pathways(input_file, species='human'):\n",
    "    dt = pd.read_table(input_file, header=None)\n",
    "    ensembl = dt[((dt[0] >= 'ENSG00000000000000') & (dt[0] <= 'ENSG00099999999999'))\n",
    "                 | ((dt[0] >= 'ENSP00000000000000') & (dt[0] <= 'ENSP00099999999999'))\n",
    "                 | ((dt[0] >= 'ENST00000000000000') & (dt[0] <= 'ENST00099999999999'))]\n",
    "\n",
    "    ensembl = ensembl.iloc[:, 0:2]\n",
    "    ensembl.columns = ['gene', 'group']\n",
    "    ensembl = pd.DataFrame(ensembl, columns=['group', 'gene'])\n",
    "    ensembl.index = range(0, ensembl.shape[0])\n",
    "\n",
    "    for i in range(0, ensembl.shape[0]):\n",
    "        ensembl.iloc[i, 1] = 'ENSG000' + ensembl.iloc[i, 1][7:18]\n",
    "\n",
    "    ensembl = ensembl.drop_duplicates()\n",
    "    ensembl.index = range(0, ensembl.shape[0])\n",
    "    return ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "692f83f4-f38d-462e-9ac2-1588c82442e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway_genes = get_gene_pathways(\"../../dataset/reactome/Ensembl2Reactome_All_Levels.txt\", species='human')\n",
    "pathway_names = '../../dataset/reactome/ReactomePathways.txt'\n",
    "relations_file_name = '../../dataset/reactome/ReactomePathwaysRelation.txt'\n",
    "root_name = [0, 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2e77658a-cc42-4807-a166-0ac83c9d3752",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Reactome():\n",
    "\n",
    "    def __init__(self, pathway_names, pathway_genes, relations_file_name, species):\n",
    "        self.pathway_names = self.load_names(pathway_names)\n",
    "        self.pathway_genes = pathway_genes\n",
    "        self.hierarchy = self.load_hierarchy(relations_file_name)\n",
    "        self.species = species\n",
    "\n",
    "    def load_names(self, pathway_names):\n",
    "        filename = pathway_names\n",
    "        df = pd.read_csv(filename, sep='\\t')\n",
    "        df.columns = ['reactome_id', 'pathway_name', 'species']\n",
    "        return df\n",
    "\n",
    "    def load_hierarchy(self, relations_file_name):\n",
    "        filename = relations_file_name\n",
    "        df = pd.read_table(filename, header=None)\n",
    "        df.columns = ['child', 'parent']\n",
    "        return df\n",
    "\n",
    "\n",
    "class ReactomeNetwork():\n",
    "\n",
    "    def __init__(self, pathway_names, pathway_genes, relations_file_name, species):\n",
    "        self.reactome = Reactome(pathway_names, pathway_genes, relations_file_name,\n",
    "                                 species)  \n",
    "        self.netx = self.get_reactome_networkx()\n",
    "\n",
    "    def get_reactome_networkx(self):\n",
    "        if hasattr(self, 'netx'):\n",
    "            return self.netx\n",
    "        hierarchy = self.reactome.hierarchy\n",
    "\n",
    "        if self.reactome.species == 'mouse':\n",
    "            abbr = 'MMU'\n",
    "        elif self.reactome.species == 'human':\n",
    "            abbr = 'HSA'\n",
    "        elif self.reactome.species == 'rat':\n",
    "            abbr = 'RNO'\n",
    "        species_hierarchy = hierarchy[hierarchy['child'].str.contains(abbr)]\n",
    "        net = nx.from_pandas_edgelist(species_hierarchy, 'child', 'parent', create_using=nx.DiGraph())\n",
    "        net.name = 'reactome'\n",
    "\n",
    "        # add root node\n",
    "        roots = [n for n, d in net.in_degree() if d == 0]\n",
    "        root_node = 'root'\n",
    "        edges = [(root_node, n) for n in roots]\n",
    "        net.add_edges_from(edges)\n",
    "\n",
    "        return net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "51d31314-0fc1-48f1-b48e-c5e312b58e68",
   "metadata": {},
   "outputs": [],
   "source": [
    "reactome_net = ReactomeNetwork(pathway_names, pathway_genes, relations_file_name, \"human\")\n",
    "original_network = reactome_net.netx\n",
    "genes_df = reactome_net.reactome.pathway_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ae2dfa21-24fa-44c4-9845-42be33d226e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "节点数量: 2581\n",
      "边的数量: 2627\n"
     ]
    }
   ],
   "source": [
    "# 获取节点数量\n",
    "num_nodes = original_network.number_of_nodes()\n",
    "\n",
    "# 获取边的数量\n",
    "num_edges = original_network.number_of_edges()\n",
    "\n",
    "# 输出节点和边的数量\n",
    "print(f\"节点数量: {num_nodes}\")\n",
    "print(f\"边的数量: {num_edges}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fcc198e0-189c-4e78-938b-22d20568b6ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 找到所有的叶子节点\n",
    "def find_leaves(graph):\n",
    "    leaves = [node for node in graph.nodes() if graph.out_degree(node) == 0]\n",
    "    return leaves\n",
    "\n",
    "# 使用 DFS 从每个节点开始，找到它最终连接到的所有叶子节点\n",
    "def find_all_leaf_nodes(graph):\n",
    "    leaves = find_leaves(graph)\n",
    "    leaf_map = {}\n",
    "\n",
    "    def dfs(node, path):\n",
    "        if node in leaves:\n",
    "            for start_node in path:\n",
    "                if start_node not in leaf_map:\n",
    "                    leaf_map[start_node] = set()\n",
    "                leaf_map[start_node].add(node)\n",
    "        else:\n",
    "            for neighbor in graph.successors(node):\n",
    "                dfs(neighbor, path + [node])\n",
    "\n",
    "    for node in graph.nodes():\n",
    "        if node not in leaf_map:  # 确保每个节点都被处理\n",
    "            dfs(node, [])\n",
    "    \n",
    "    # 确保每个叶子节点自己在自己的映射中\n",
    "    for leaf in leaves:\n",
    "        if leaf not in leaf_map:\n",
    "            leaf_map[leaf] = set()\n",
    "        leaf_map[leaf].add(leaf)\n",
    "\n",
    "    return leaf_map\n",
    "    \n",
    "def add_genes_to_network(network, genes_df, leaf_map):\n",
    "    for index, row in genes_df.iterrows():\n",
    "        group = row['group']  # Pathway name\n",
    "        gene = row['gene']    # Gene name\n",
    "        \n",
    "       # Add the gene as a node if it doesn't already exist\n",
    "        if not network.has_node(gene):\n",
    "            network.add_node(gene, type='gene')  # You can add more attributes if needed\n",
    "        \n",
    "        # Find leaf nodes for the given pathway using the original leaf nodes map\n",
    "        if group in leaf_map:\n",
    "            leaf_nodes = leaf_map[group]\n",
    "            # Add an edge from each leaf node to the gene\n",
    "            for leaf in leaf_nodes:\n",
    "                network.add_edge(leaf, gene)\n",
    "    \n",
    "    return network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "acd8c7e3-8e5d-4872-83d8-bdf8791a4f08",
   "metadata": {},
   "outputs": [],
   "source": [
    "leaf_map = find_all_leaf_nodes(original_network)\n",
    "final_network = add_genes_to_network(original_network, genes_df, leaf_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "746161ee-7269-449a-9789-17681d82a9fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "节点数量: 65195\n",
      "边的数量: 18166335\n"
     ]
    }
   ],
   "source": [
    "# 获取节点数量\n",
    "num_nodes = final_network.number_of_nodes()\n",
    "\n",
    "# 获取边的数量\n",
    "num_edges = final_network.number_of_edges()\n",
    "\n",
    "# 输出节点和边的数量\n",
    "print(f\"节点数量: {num_nodes}\")\n",
    "print(f\"边的数量: {num_edges}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "83f68930-c136-4f8a-bfaf-9af789d18d30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def add_duplicated_edges(G, node, n_levels):\n",
    "#     edges = []\n",
    "#     source = node\n",
    "#     for l in range(n_levels):\n",
    "#         target = node + '_copy' + str(l + 1)\n",
    "#         edge = (source, target)\n",
    "#         source = target\n",
    "#         edges.append(edge)\n",
    "\n",
    "#     G.add_edges_from(edges)\n",
    "#     return G, target\n",
    "\n",
    "\n",
    "# def add_gene_edges(G, node, pathways, genes_df):\n",
    "#     genes = []\n",
    "#     if type(pathways) == str:\n",
    "#         genes = genes + list(genes_df[genes_df['group'] == pathways]['gene'])\n",
    "#     else:\n",
    "#         for i in range(len(pathways)):\n",
    "#             genes = genes + list(genes_df[genes_df['group'] == pathways[i]]['gene'])\n",
    "#     genes = list(set(genes))\n",
    "#     edges = []\n",
    "#     source = node\n",
    "#     for target in genes:\n",
    "#         edge = (source, target)\n",
    "#         edges.append(edge)\n",
    "#     G.add_edges_from(edges)\n",
    "#     return G\n",
    "\n",
    "\n",
    "# def get_nodes_at_level(net, distance):\n",
    "#     nodes = set(nx.ego_graph(net, 'root', radius=distance))  # get all nodes within distance around the query node\n",
    "#     if distance >= 1.:\n",
    "#         nodes -= set(nx.ego_graph(net, 'root',\n",
    "#                                   radius=distance - 1))  # remove nodes that are not at the specified distance but closer\n",
    "#     return list(nodes)\n",
    "\n",
    "\n",
    "# def gene_mapping(gene, df):\n",
    "#     inter_gene = list(set(gene) & set(df['gene']))\n",
    "#     genedict = {}\n",
    "#     genelist = [df.iloc[0, 0]]\n",
    "#     genedict[df.iloc[0, 1]] = genelist\n",
    "#     for i in range(1, df.shape[0]):\n",
    "#         if df.iloc[i, 1] == df.iloc[i - 1, 1]:\n",
    "#             genelist.append(df.iloc[i, 0])\n",
    "#         else:\n",
    "#             genedict[df.iloc[i - 1, 1]] = genelist\n",
    "#             genelist = [df.iloc[i, 0]]\n",
    "#     mappingdf = pd.DataFrame(data=None, columns=['group', 'gene'])\n",
    "#     for j in range(len(inter_gene)):\n",
    "#         mappingdf_iter = {'group': genedict[inter_gene[j]],\n",
    "#                           'gene': [inter_gene[j]] * len(genedict[inter_gene[j]])}\n",
    "#         mappingdf_iter = pd.DataFrame(mappingdf_iter)\n",
    "#         mappingdf = pd.concat([mappingdf, mappingdf_iter])\n",
    "#     return mappingdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9b7422d5-4608-428c-88d4-9be3419a2026",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_gene_in = data.columns.tolist()\n",
    "# species = 'human'\n",
    "# n_hidden = 3\n",
    "\n",
    "# reactome_net = ReactomeNetwork(pathway_names, pathway_genes, relations_file_name, species)\n",
    "# genes_df = reactome_net.reactome.pathway_genes\n",
    "# genes_df = gene_mapping(data_gene_in, genes_df)\n",
    "\n",
    "# original_network = reactome_net.netx\n",
    "# original_terminal_nodes = [n for n, d in original_network.out_degree() if d == 0]\n",
    "# in_genes_df = [False for x in range(0, len(original_terminal_nodes))]\n",
    "# while in_genes_df.count(False) > 0:\n",
    "#     in_genes_df = [False for x in range(0, len(original_terminal_nodes))]\n",
    "#     for i in range(len(original_terminal_nodes)):\n",
    "#         if original_terminal_nodes[i] in genes_df['group'].to_list():\n",
    "#             in_genes_df[i] = True\n",
    "#     for i in range(len(original_terminal_nodes)):\n",
    "#         if in_genes_df[i] == False:\n",
    "#             original_network.remove_node(original_terminal_nodes[i])\n",
    "#     original_terminal_nodes = [n for n, d in original_network.out_degree() if d == 0]\n",
    "\n",
    "# sub_graph = nx.ego_graph(original_network, 'root',\n",
    "#                          radius=n_hidden)  # subgraph of neighbors centered at node \"root\" <= a given radius (n_level).\n",
    "# sub_terminal_nodes = [n for n, d in sub_graph.out_degree() if d == 0]\n",
    "# for node in sub_terminal_nodes:\n",
    "#     distance = len(nx.shortest_path(sub_graph, source='root',\n",
    "#                                     target=node))  # len of distance: num of nodes in the shortest path\n",
    "#     if (distance == n_hidden + 1) & (node not in original_terminal_nodes):\n",
    "#         part_graph = nx.ego_graph(original_network, node, radius=100)\n",
    "#         corresponding_terminal_nodes = [n for n, d in part_graph.out_degree() if d == 0]\n",
    "#         sub_graph = add_gene_edges(sub_graph, node, corresponding_terminal_nodes, genes_df)\n",
    "#     elif (distance == n_hidden + 1) & (node in original_terminal_nodes):\n",
    "#         corresponding_terminal_nodes = node\n",
    "#         sub_graph = add_gene_edges(sub_graph, node, corresponding_terminal_nodes, genes_df)\n",
    "#     elif distance <= n_hidden:\n",
    "#         diff = n_hidden - distance + 1\n",
    "#         sub_graph, copy_node = add_duplicated_edges(sub_graph, node, diff)\n",
    "#         sub_graph = add_gene_edges(sub_graph, copy_node, node, genes_df)\n",
    "# final_network = sub_graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e5d050cf-b591-4066-8ca8-216db9e0fe54",
   "metadata": {},
   "outputs": [],
   "source": [
    "level_dict = {}\n",
    "roots = [node for node in final_network.nodes if final_network.in_degree(node) == 0]\n",
    "for root in roots:\n",
    "    path_lengths = nx.single_source_shortest_path_length(final_network, root)\n",
    "    for node, length in path_lengths.items():\n",
    "        if node in level_dict:\n",
    "            level_dict[node] = min(level_dict[node], length)\n",
    "        else:\n",
    "            level_dict[node] = length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d03a37f5-bd29-41e8-9508-22251b68fb10",
   "metadata": {},
   "outputs": [],
   "source": [
    "levels = collections.defaultdict(list)\n",
    "for node, level in level_dict.items():\n",
    "    if not node.startswith(\"ENSG\"):\n",
    "        levels[level].append(node)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e01e53fd-2db1-4135-98e1-706009c3249a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_gene['ensembl'] = mut_gene[0].map(gene_info.set_index('From')['To'])\n",
    "mut_gene.columns = ['gene name', 'ensembl']\n",
    "mut_gene = mut_gene.iloc[0:10, :]\n",
    "mut_gene = mut_gene.sort_values(by='gene name')\n",
    "mut_gene.index = range(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8d097cde-8136-4536-b758-7a2a80c83918",
   "metadata": {},
   "outputs": [],
   "source": [
    "p1 = pd.read_csv(\"../../result/supervised/pathway/pathways_BRAF.csv\", index_col = 0)\n",
    "p2 = pd.read_csv(\"../../result/supervised/pathway/pathways_CDH3.csv\", index_col = 0)\n",
    "p3 = pd.read_csv(\"../../result/supervised/pathway/pathways_HRAS.csv\", index_col = 0)\n",
    "p4 = pd.read_csv(\"../../result/supervised/pathway/pathways_KRAS.csv\", index_col = 0)\n",
    "p5 = pd.read_csv(\"../../result/supervised/pathway/pathways_NRAS.csv\", index_col = 0)\n",
    "p6 = pd.read_csv(\"../../result/supervised/pathway/pathways_PIK3CA.csv\", index_col = 0)\n",
    "p7 = pd.read_csv(\"../../result/supervised/pathway/pathways_SCN5A.csv\", index_col = 0)\n",
    "p8 = pd.read_csv(\"../../result/supervised/pathway/pathways_SERPINH1.csv\", index_col = 0)\n",
    "p9 = pd.read_csv(\"../../result/supervised/pathway/pathways_TP53.csv\", index_col = 0)\n",
    "p10 = pd.read_csv(\"../../result/supervised/pathway/pathways_VDR.csv\", index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "77a4b361-7e0a-4a79-b0af-937ee068b801",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway = [p1, p2, p3, p4, p5, p6, p7, p8, p9, p10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f34c702a-8428-4e43-ae14-fe6e0de7c79d",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_length = {}\n",
    "for i in range(0, 10):\n",
    "    path_length[mut_gene.iloc[i, 0]] = pd.DataFrame(columns=['experimental group', 'control group'])\n",
    "    for p in pathway[i]['pathway']:\n",
    "        try:\n",
    "            exp_distance = nx.shortest_path_length(final_network, source=p, target=mut_gene.iloc[i, 1])\n",
    "        except nx.NetworkXNoPath:\n",
    "            exp_distance = float('inf') \n",
    "        rand_pathway = levels[level_dict[p]][random.randint(0, len(levels[level_dict[p]]) - 1)]\n",
    "        try:\n",
    "            ctrl_distance = nx.shortest_path_length(final_network, source=rand_pathway, target=mut_gene.iloc[i, 1])\n",
    "        except nx.NetworkXNoPath:\n",
    "            ctrl_distance = float('inf') \n",
    "        path_length[mut_gene.iloc[i, 0]].loc[p] = [exp_distance, ctrl_distance]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "00aa8405-09da-481a-bc70-e2ff71e0872e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k in path_length.keys():\n",
    "    path_length[k].to_csv(k + \"_path_length.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "55bc5650-cc58-40be-86e1-022d0f286530",
   "metadata": {},
   "outputs": [],
   "source": [
    "undirected_network = final_network.to_undirected()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "55745b95-d6aa-4907-8d03-66340db1870b",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_adj_all = pd.DataFrame(index=mut_gene['gene name'], columns=range(5))\n",
    "correlation_all = pd.DataFrame(index=mut_gene['gene name'], columns=range(5))\n",
    "path_length = {}\n",
    "for i in range(0, 10):\n",
    "    p_value, correlation = [], []\n",
    "    for times in range(5):\n",
    "        path_length[mut_gene.iloc[i, 0]] = pd.DataFrame(columns=['experimental group', 'control group'])\n",
    "        for p in pathway[i]['pathway']:\n",
    "            exp_distance = nx.shortest_path_length(undirected_network, source=p, target=mut_gene.iloc[i, 1])\n",
    "            rand_pathway = levels[level_dict[p]][random.randint(0, len(levels[level_dict[p]]) - 1)]\n",
    "            ctrl_distance = nx.shortest_path_length(undirected_network, source=rand_pathway, target=mut_gene.iloc[i, 1])\n",
    "            path_length[mut_gene.iloc[i, 0]].loc[p] = [exp_distance, ctrl_distance]\n",
    "        _, pval = wilcoxon(path_length[mut_gene.iloc[i, 0]]['experimental group'], path_length[mut_gene.iloc[i, 0]]['control group'])\n",
    "        corr, _ = pearsonr((path_length[mut_gene.iloc[i, 0]]['experimental group'] - path_length[mut_gene.iloc[i, 0]]['control group']) / path_length[mut_gene.iloc[i, 0]]['experimental group'], range(1, len(pathway[i]['pathway']) + 1))\n",
    "        p_value.append(pval)\n",
    "        correlation.append(corr)\n",
    "    _, p_value_adj, _, _ = multipletests(p_value, alpha=0.05, method='fdr_bh')\n",
    "    p_adj_all.loc[mut_gene.iloc[i, 0]] = p_value_adj\n",
    "    correlation_all.loc[mut_gene.iloc[i, 0]] = correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3030b0b9-30d6-48a6-ae08-e99677351d30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BRAF</th>\n",
       "      <td>0.125047</td>\n",
       "      <td>0.46353</td>\n",
       "      <td>0.015377</td>\n",
       "      <td>0.799195</td>\n",
       "      <td>0.388141</td>\n",
       "      <td>0.13141</td>\n",
       "      <td>0.707594</td>\n",
       "      <td>0.027831</td>\n",
       "      <td>0.0345</td>\n",
       "      <td>0.536593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH3</th>\n",
       "      <td>0.003775</td>\n",
       "      <td>0.003172</td>\n",
       "      <td>0.006747</td>\n",
       "      <td>0.572606</td>\n",
       "      <td>0.133991</td>\n",
       "      <td>0.593869</td>\n",
       "      <td>0.012039</td>\n",
       "      <td>0.003111</td>\n",
       "      <td>0.001839</td>\n",
       "      <td>0.431719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HRAS</th>\n",
       "      <td>0.164354</td>\n",
       "      <td>0.177398</td>\n",
       "      <td>0.037949</td>\n",
       "      <td>0.204887</td>\n",
       "      <td>0.56368</td>\n",
       "      <td>0.018658</td>\n",
       "      <td>0.058079</td>\n",
       "      <td>0.013519</td>\n",
       "      <td>0.222223</td>\n",
       "      <td>0.06227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRAS</th>\n",
       "      <td>0.051801</td>\n",
       "      <td>0.000948</td>\n",
       "      <td>0.01543</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.026087</td>\n",
       "      <td>0.004961</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.000936</td>\n",
       "      <td>0.003473</td>\n",
       "      <td>0.000344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NRAS</th>\n",
       "      <td>0.252034</td>\n",
       "      <td>0.613662</td>\n",
       "      <td>0.460742</td>\n",
       "      <td>0.097022</td>\n",
       "      <td>0.044591</td>\n",
       "      <td>0.034364</td>\n",
       "      <td>0.08511</td>\n",
       "      <td>0.373363</td>\n",
       "      <td>0.152147</td>\n",
       "      <td>0.003104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PIK3CA</th>\n",
       "      <td>0.053721</td>\n",
       "      <td>0.341248</td>\n",
       "      <td>0.160946</td>\n",
       "      <td>0.637825</td>\n",
       "      <td>0.028501</td>\n",
       "      <td>0.000982</td>\n",
       "      <td>0.106467</td>\n",
       "      <td>0.432259</td>\n",
       "      <td>0.725663</td>\n",
       "      <td>0.179312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCN5A</th>\n",
       "      <td>0.036126</td>\n",
       "      <td>0.03094</td>\n",
       "      <td>0.038611</td>\n",
       "      <td>0.075553</td>\n",
       "      <td>0.009233</td>\n",
       "      <td>0.078602</td>\n",
       "      <td>0.099907</td>\n",
       "      <td>0.022456</td>\n",
       "      <td>0.028034</td>\n",
       "      <td>0.363139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SERPINH1</th>\n",
       "      <td>0.862406</td>\n",
       "      <td>0.927034</td>\n",
       "      <td>0.274964</td>\n",
       "      <td>0.925459</td>\n",
       "      <td>0.335521</td>\n",
       "      <td>0.582693</td>\n",
       "      <td>0.903071</td>\n",
       "      <td>0.491551</td>\n",
       "      <td>0.45699</td>\n",
       "      <td>0.259713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53</th>\n",
       "      <td>0.95177</td>\n",
       "      <td>0.907733</td>\n",
       "      <td>0.250596</td>\n",
       "      <td>0.284301</td>\n",
       "      <td>0.356166</td>\n",
       "      <td>0.361591</td>\n",
       "      <td>0.584104</td>\n",
       "      <td>0.807429</td>\n",
       "      <td>0.43743</td>\n",
       "      <td>0.257568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VDR</th>\n",
       "      <td>0.109771</td>\n",
       "      <td>0.013385</td>\n",
       "      <td>0.043453</td>\n",
       "      <td>0.016108</td>\n",
       "      <td>0.030722</td>\n",
       "      <td>0.016923</td>\n",
       "      <td>0.027754</td>\n",
       "      <td>0.029565</td>\n",
       "      <td>0.004604</td>\n",
       "      <td>0.002066</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0         1         2         3         4         5  \\\n",
       "gene name                                                               \n",
       "BRAF       0.125047   0.46353  0.015377  0.799195  0.388141   0.13141   \n",
       "CDH3       0.003775  0.003172  0.006747  0.572606  0.133991  0.593869   \n",
       "HRAS       0.164354  0.177398  0.037949  0.204887   0.56368  0.018658   \n",
       "KRAS       0.051801  0.000948   0.01543  0.000108  0.026087  0.004961   \n",
       "NRAS       0.252034  0.613662  0.460742  0.097022  0.044591  0.034364   \n",
       "PIK3CA     0.053721  0.341248  0.160946  0.637825  0.028501  0.000982   \n",
       "SCN5A      0.036126   0.03094  0.038611  0.075553  0.009233  0.078602   \n",
       "SERPINH1   0.862406  0.927034  0.274964  0.925459  0.335521  0.582693   \n",
       "TP53        0.95177  0.907733  0.250596  0.284301  0.356166  0.361591   \n",
       "VDR        0.109771  0.013385  0.043453  0.016108  0.030722  0.016923   \n",
       "\n",
       "                  6         7         8         9  \n",
       "gene name                                          \n",
       "BRAF       0.707594  0.027831    0.0345  0.536593  \n",
       "CDH3       0.012039  0.003111  0.001839  0.431719  \n",
       "HRAS       0.058079  0.013519  0.222223   0.06227  \n",
       "KRAS       0.000447  0.000936  0.003473  0.000344  \n",
       "NRAS        0.08511  0.373363  0.152147  0.003104  \n",
       "PIK3CA     0.106467  0.432259  0.725663  0.179312  \n",
       "SCN5A      0.099907  0.022456  0.028034  0.363139  \n",
       "SERPINH1   0.903071  0.491551   0.45699  0.259713  \n",
       "TP53       0.584104  0.807429   0.43743  0.257568  \n",
       "VDR        0.027754  0.029565  0.004604  0.002066  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_adj_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "12fed0d4-177d-4c9f-a316-706c482712b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BRAF</th>\n",
       "      <td>0.005815</td>\n",
       "      <td>0.04799</td>\n",
       "      <td>0.073674</td>\n",
       "      <td>-0.028375</td>\n",
       "      <td>0.042578</td>\n",
       "      <td>0.067445</td>\n",
       "      <td>0.100688</td>\n",
       "      <td>0.037487</td>\n",
       "      <td>0.045336</td>\n",
       "      <td>0.007804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH3</th>\n",
       "      <td>-0.014054</td>\n",
       "      <td>0.025068</td>\n",
       "      <td>-0.014793</td>\n",
       "      <td>-0.095549</td>\n",
       "      <td>-0.034256</td>\n",
       "      <td>0.093862</td>\n",
       "      <td>-0.013984</td>\n",
       "      <td>-0.063431</td>\n",
       "      <td>-0.010184</td>\n",
       "      <td>-0.06519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HRAS</th>\n",
       "      <td>-0.043166</td>\n",
       "      <td>-0.008696</td>\n",
       "      <td>-0.073381</td>\n",
       "      <td>0.072629</td>\n",
       "      <td>-0.012993</td>\n",
       "      <td>0.003121</td>\n",
       "      <td>0.007234</td>\n",
       "      <td>0.001697</td>\n",
       "      <td>-0.00935</td>\n",
       "      <td>0.06805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRAS</th>\n",
       "      <td>-0.035614</td>\n",
       "      <td>-0.133171</td>\n",
       "      <td>-0.021351</td>\n",
       "      <td>-0.114536</td>\n",
       "      <td>-0.040312</td>\n",
       "      <td>0.048684</td>\n",
       "      <td>-0.133687</td>\n",
       "      <td>-0.055568</td>\n",
       "      <td>0.01439</td>\n",
       "      <td>-0.015135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NRAS</th>\n",
       "      <td>0.058169</td>\n",
       "      <td>0.02669</td>\n",
       "      <td>-0.054718</td>\n",
       "      <td>0.024983</td>\n",
       "      <td>0.144563</td>\n",
       "      <td>0.001725</td>\n",
       "      <td>0.07261</td>\n",
       "      <td>0.085081</td>\n",
       "      <td>0.137342</td>\n",
       "      <td>0.092308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PIK3CA</th>\n",
       "      <td>0.171811</td>\n",
       "      <td>0.023636</td>\n",
       "      <td>-0.00738</td>\n",
       "      <td>0.129893</td>\n",
       "      <td>0.195845</td>\n",
       "      <td>0.124946</td>\n",
       "      <td>0.19581</td>\n",
       "      <td>0.059281</td>\n",
       "      <td>0.213509</td>\n",
       "      <td>0.116204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCN5A</th>\n",
       "      <td>-0.097598</td>\n",
       "      <td>-0.074352</td>\n",
       "      <td>-0.135635</td>\n",
       "      <td>-0.085725</td>\n",
       "      <td>-0.163713</td>\n",
       "      <td>-0.062927</td>\n",
       "      <td>-0.163853</td>\n",
       "      <td>-0.087931</td>\n",
       "      <td>-0.164325</td>\n",
       "      <td>-0.114969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SERPINH1</th>\n",
       "      <td>-0.159285</td>\n",
       "      <td>-0.167137</td>\n",
       "      <td>-0.126108</td>\n",
       "      <td>-0.184493</td>\n",
       "      <td>-0.17585</td>\n",
       "      <td>-0.158564</td>\n",
       "      <td>-0.1864</td>\n",
       "      <td>-0.141712</td>\n",
       "      <td>-0.123972</td>\n",
       "      <td>-0.146011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53</th>\n",
       "      <td>0.024165</td>\n",
       "      <td>-0.043577</td>\n",
       "      <td>-0.056507</td>\n",
       "      <td>-0.133842</td>\n",
       "      <td>0.019625</td>\n",
       "      <td>0.047115</td>\n",
       "      <td>-0.036224</td>\n",
       "      <td>-0.038402</td>\n",
       "      <td>0.045475</td>\n",
       "      <td>-0.11877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VDR</th>\n",
       "      <td>-0.096051</td>\n",
       "      <td>-0.102953</td>\n",
       "      <td>-0.090849</td>\n",
       "      <td>0.003287</td>\n",
       "      <td>-0.000781</td>\n",
       "      <td>-0.070694</td>\n",
       "      <td>-0.030342</td>\n",
       "      <td>-0.074312</td>\n",
       "      <td>-0.075124</td>\n",
       "      <td>-0.015535</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0         1         2         3         4         5  \\\n",
       "gene name                                                               \n",
       "BRAF       0.005815   0.04799  0.073674 -0.028375  0.042578  0.067445   \n",
       "CDH3      -0.014054  0.025068 -0.014793 -0.095549 -0.034256  0.093862   \n",
       "HRAS      -0.043166 -0.008696 -0.073381  0.072629 -0.012993  0.003121   \n",
       "KRAS      -0.035614 -0.133171 -0.021351 -0.114536 -0.040312  0.048684   \n",
       "NRAS       0.058169   0.02669 -0.054718  0.024983  0.144563  0.001725   \n",
       "PIK3CA     0.171811  0.023636  -0.00738  0.129893  0.195845  0.124946   \n",
       "SCN5A     -0.097598 -0.074352 -0.135635 -0.085725 -0.163713 -0.062927   \n",
       "SERPINH1  -0.159285 -0.167137 -0.126108 -0.184493  -0.17585 -0.158564   \n",
       "TP53       0.024165 -0.043577 -0.056507 -0.133842  0.019625  0.047115   \n",
       "VDR       -0.096051 -0.102953 -0.090849  0.003287 -0.000781 -0.070694   \n",
       "\n",
       "                  6         7         8         9  \n",
       "gene name                                          \n",
       "BRAF       0.100688  0.037487  0.045336  0.007804  \n",
       "CDH3      -0.013984 -0.063431 -0.010184  -0.06519  \n",
       "HRAS       0.007234  0.001697  -0.00935   0.06805  \n",
       "KRAS      -0.133687 -0.055568   0.01439 -0.015135  \n",
       "NRAS        0.07261  0.085081  0.137342  0.092308  \n",
       "PIK3CA      0.19581  0.059281  0.213509  0.116204  \n",
       "SCN5A     -0.163853 -0.087931 -0.164325 -0.114969  \n",
       "SERPINH1    -0.1864 -0.141712 -0.123972 -0.146011  \n",
       "TP53      -0.036224 -0.038402  0.045475  -0.11877  \n",
       "VDR       -0.030342 -0.074312 -0.075124 -0.015535  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "correlation_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "09c4aa31-ccb1-4fd1-90b3-a046b5319bad",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_adj_all.to_csv(\"p_adj.csv\")\n",
    "correlation_all.to_csv(\"corr.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.2",
   "language": "python",
   "name": "py392"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
